src/theia/sfm/pose/four_point_focal_length_radial_distortion.cc

```diff
@@ -0,0 +1,632 @@
+    };
+    Eigen::MatrixXd C = Eigen::MatrixXd::Zero(40,50);
+    for (int i = 0; i < C_ind_r.size(); ++i)
+        C(C_ind_r[i], C_ind_c[i]) = coeffs1(coeffs_ind_r[i], 0);
+
+    std::vector<int> b_ind_r = { 0  ,   1  ,   2 ,    3 ,    4   ,  5  ,   6 };
+    std::vector<int> b_ind_c = { 30 ,   31 ,   32  ,  33,    34,    35 ,   36 };
+
+    Eigen::MatrixXd b = Eigen::MatrixXd::Zero(7,37);
+    for (int i = 0; i < b_ind_r.size(); ++i)
+        b(b_ind_r[i], b_ind_c[i]) = -1.0;
+
+    Eigen::MatrixXd C0 = Eigen::MatrixXd::Zero(40, 37);
+    C0 = C.leftCols<37>();
+    Eigen::MatrixXd	C1 = Eigen::MatrixXd::Zero(40, 13);
+    C1 = C.rightCols<13>();
+
+    Eigen::MatrixXd alpha = C0.transpose().fullPivLu().solve(b.transpose());
+    Eigen::MatrixXd RR = Eigen::MatrixXd::Zero(20,13);
+    RR.block<7, 13>(0, 0) = alpha.transpose() * C1;
+    RR.block<13, 13>(7, 0) = Eigen::MatrixXd::Identity(13, 13);
+    std::vector<int> AM_ind = { 17  ,   9 ,    0  ,  11  ,   1  ,  13  ,   2  ,  16  ,   3  ,   4 ,   18  ,   5  ,   6 };
+    Eigen::MatrixXd AM = Eigen::MatrixXd::Zero(13,13);
+    for (int i = 0; i < AM_ind.size(); ++i)
+        AM.row(i) = RR.row(AM_ind[i]);
+
+    Eigen::EigenSolver<Eigen::MatrixXd> eig;
+    eig.compute(AM);
+
+    Eigen::ArrayXcd D = eig.eigenvalues();
+    Eigen::ArrayXXcd V = eig.eigenvectors();
+    V = (V / V.row(0).replicate(13, 1)).eval();
+    sols.row(0) = V.row(5);
+    sols.row(1) = V.row(7);
+    sols.row(2) = D.transpose();
+    sols.row(3) = V.row(1);
+    sols.row(4) = V.row(3);
+}
+
+
+bool FourPointsPoseFocalLengthRadialDistortion(
+    const Eigen::Matrix<double, 3, 4>& feature_vectors,
+    const Eigen::Matrix4d& world_points,
+    std::vector<Eigen::Matrix3d>& rotations,
+    std::vector<Eigen::Vector3d>& translations,
+    std::vector<double>& radial_distortions,
+    std::vector<double>& focal_lengths)
+{
+    Eigen::Vector4d d = feature_vectors.topRows<2>().array().pow(2).colwise().sum();
+
+    Eigen::Vector3d t0 = world_points.topRows<3>().rowwise().mean();
+    Eigen::Matrix<double, 3, 4> t0_mat;
+    t0_mat.col(0) = t0; t0_mat.col(1) = t0; t0_mat.col(2) = t0; t0_mat.col(3) = t0;
+    Eigen::Matrix<double, 4, 4> U = world_points;
+    U.topRows<3>() -= t0_mat;
+
+    Eigen::JacobiSVD < Eigen::MatrixXd > svd(U.topRows<3>(), Eigen::ComputeFullV | Eigen::ComputeFullU);
+    Eigen::Matrix3d R0 = svd.matrixU();
+
+    if (sgn(R0.determ
```

Comment:
Can you push this logic into the Solve method, and have Solve return a std::vector of Eigen::Matrix<double, 5, 1>? That will simplify the logic here and below a bit